xdsl/dialects/bufferization.py

```diff
@@ -14,14 +14,61 @@
 from xdsl.irdl import (
     AnyOf,
     AttrSizedOperandSegments,
+    ConstraintContext,
     IRDLOperation,
+    VarConstraint,
     irdl_op_definition,
     operand_def,
     opt_operand_def,
     opt_prop_def,
     result_def,
     var_operand_def,
 )
+from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.hints import isa
+
+
+class TensorMemrefInferenceConstraint(VarConstraint[Attribute]):
+    """
+    Constraint to infer tensor shapes from memref shapes.
```

Comment:
Could you please add a description of how the mechanism below works for inference and verification?